File: monotonic-queue/sliding-window-maximum.py
class MyQue:
    def __init__(self):
        self.que = deque()
    def push(self, value: int):
        while len(self.que)!=0 and self.que[-1]<value:
            self.que.pop()
        self.que.append(value)
    def pop(self, value: int):
        if len(self.que)!=0 and self.que[0]==value:
            self.que.popleft()
    def front(self):
        return self.que[0]


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        result = []
        que = MyQue()

        for i in range(k):
            que.push(nums[i])
        
        result.append(que.front())

        for i in range(k, len(nums)):
            que.pop(nums[i-k])
            que.push(nums[i])
            result.append(que.front())
        
        return result

File: sliding-window/sliding-window-maximum.cpp
class Solution {
private:
    class MyQue{
    private:
        deque<int> que;
    public:
        MyQue(){

        }

        void push(int value){
            while (!que.empty() && que.back() <value){
                que.pop_back();
            }
            que.push_back(value);
        }

        void pop(int value){
            if (!que.empty() && que.front() == value){
                que.pop_front();
            }
        }

        int front(){
            return que.front();
        }
    };
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> result;
        MyQue que;

        for (int i = 0; i<k; i++){
            que.push(nums[i]);
        }
        result.push_back(que.front());

        for (int i=k; i<nums.size();i++){
            que.pop(nums[i-k]);
            que.push(nums[i]);
            result.push_back(que.front());
        }
        return result;
    }
};